kernel/posix/pthread.c

```diff
@@ -0,0 +1,351 @@
+		*policy = SCHED_FIFO;
+		prio = -1 * (z_prio + 1);
+	} else {
+		*policy = SCHED_RR;
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - z_prio);
+	}
+
+	return prio;
+}
+
+static inline s32_t posix_to_zephyr_priority(u32_t priority, int policy)
+{
+	s32_t prio;
+
+	if (policy == SCHED_FIFO) {
+		/* Zephyr COOP priority starts from -1 */
+		prio =  -1 * (priority + 1);
+	} else {
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+	}
+
+	return prio;
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (is_posix_prio_valid(priority, attr->schedpolicy) == false) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	attr->priority = priority;
+	return 0;
+}
+
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+	if (stackaddr == NULL) {
+		errno = EACCES;
+		return -1;
+	}
+
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	atomic_val_t thread_index = 0;
+	s32_t prio;
+	pthread_mutexattr_t att;
+	pthread_condattr_t cond_attr;
+
+	if (!attr || !attr->stack) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	thread_index = atomic_inc(&thread_count);
+
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		pthread[thread_index].cancel_state = (1 << PTHREAD_CANCEL_POS) &
+						      (attr->options);
+		pthread[thread_index].state = attr->detachstate;
+		pthread[thread_index].cancel_pending = 0;
+
+		pthread[thread_index].state_lock.sem =
+					&pthread[thread_index].state_lock_sem;
+		pthread_mutex_init(&pthread[thread_index].state_lock, &att);
+		pthread[thread_index].cancel_lock.sem =
+					&pthread[thread_index].cancel_lock_sem;
+		pthread_mutex_init(&pthread[thread_index].cancel_lock, &att);
+		pthread_cond_init(&pthread[thread_index].state_cond,
+				  &cond_attr);
+		*newthread = (pthread_t)
+				k_thread_create(&pthread[thread_index].thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->options,
+						attr->delayedstart);
+
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		errno = EAGAIN;
+		return -1;
+	}
+
+	return 0;
+}
+
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	unsigned int key;
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE &&
+	    state != PTHREAD_CANCEL_DISABLE) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread->cancel_state = state;
+
+	key = irq_lock();
```

Comment:
I reckon that some of the pthread stuff will eventually be system calls. User mode code can't lock interrupts.

Reply:
Addressed. I removed irq_lock and using cancel_state mutex for this.